ua-snap/cmip6-utils
# Add regridding QC

This PR adds QC to the regridding pipeline (closes #41 ).

The visual QC notebook has been updated to run via `papermill`, and a separate `qc.py` script has been added. The `qc.py` script parses slurm output files for messages stored during regridding, checks that all expected regridded files exist, and tests that the regridded files open and have values that are within the source file range.

An additional `run_qc.py` script was added in order to submit both of the above QC processes as `sbatch` jobs, so that we don't run the entire QC process from the login node.

**TO TEST**:

Let's limit the testing to just a few batch files to speed things up. Make sure your scratch directory is empty from previous runs (or make a brand new one for this test) and manually create the  `{scratch_directory}/cmip6_regridding/regrid_batch` folder. Drop these files in that folder (or create them with `vim` and copy/paste the contents):

[batch_CNRM-CM6-1-HR_ssp245_gr0_0.txt](https://github.com/ua-snap/cmip6-utils/files/14671673/batch_CNRM-CM6-1-HR_ssp245_gr0_0.txt)
[batch_TaiESM1_ssp245_gr0_0.txt](https://github.com/ua-snap/cmip6-utils/files/14671674/batch_TaiESM1_ssp245_gr0_0.txt)

Now run the regridding flow from Prefect. Use the `qc` branch of the cmip6-utils repo. Set `no_clobber='false'`, `generate_batch_files='false'`, and `vars='global'`. This will generate the slurm files, the regridded files, plus the new QC directory and files.

Check the QC outputs in `{output_directory}/qc`. There will be slurm `*.out` files, a `qc_error.txt` file, and the QC notebook's `visual_qc_out.ipynb` and `visual_qc_out.html` output files. You should see some printed statements in the slurm output files, the text file should be empty, and the QC notebook should be produced without errors. (Like with the indicators QC notebook, you'll want to copy the `visual_qc_out.html` file to your local machine to view it.)

Now we will introduce some bugs to test the QC process:

- Add a bad filename to one of the batch text files; this will fail during regridding, and should appear in the subsequent QC error outputs. I did this by just adding a bogus *.nc file as the last line, something like `this_is_a_bad_file.nc`.
- Corrupt a file in the regridded output; this should fail to open during QC and should appear in the subsequent QC error outputs. I just deleted one of the regridded files, then used `vim` to create an empty file with the same name. The idea is that `xarray` will try and fail to open, and this file will be caught.
- Introduce values outside range; open any regridded file, and multiply the values by 10,000 to create an egregious error; this will be noticed during QC and should appear in the subsequent QC error outputs. In a scratch notebook I did the following:
 ```
import xarray as xr
fp = '/center1/CMIP6/jdpaul3/regrid/cmip6_regridding/regrid/CNRM-CM6-1-HR/ssp245/Amon/ta/ta_Amon_CNRM-CM6-1-HR_ssp245_regrid_208901-208912.nc'
with xr.open_dataset(fp) as ds:
    ds_ = ds*10000
ds_.to_netcdf(fp, mode='w')
```

After making the 3 bugs above, run the Prefect flow again but set `no_clobber='true'`. The QC files should show errors related to all 3 bugs, plus some messaging about not overwriting files.

Delete the files in the `{output_directory}/qc` folder, then simulate one more bug:

- Remove one or more regridded files; this should be noticed during QC and should appear in the QC error outputs. I just opened up a random folder in the regridded outputs and temporarily moved a file to my home directory.

We have to test this bug a different way, since running the Prefect flow would just re-create this file. So instead of using Prefect, we can just use the `run_qc.py` script from the command line like so (making sure you are in the `{scratch_directory}/cmip6-utils/regridding` directory, and replacing the file paths with your local files):

```
python run_qc.py --qc_script '/center1/CMIP6/jdpaul3/regrid/cmip6-utils/regridding/qc.py' --visual_qc_notebook '/center
1/CMIP6/jdpaul3/regrid/cmip6-utils/regridding/qc.ipynb' --conda_init_script '/center1/CMIP6/jdpaul3/regrid/cmip6-utils/regridding/conda_init.sh' --cmip6_directory '/beegfs/
CMIP6/arctic-cmip6/CMIP6' --output_directory '/center1/CMIP6/jdpaul3/regrid/cmip6_regridding' --repo_regridding_directory '/center1/CMIP6/jdpaul3/regrid/cmip6-utils/regridd
ing' --slurm_email 'jdpaul3@alaska.edu' --vars 'clt evspsbl hfls hfss hus huss orog pr prsn ps psl rlds rls rsds rss sfcWind sfcWindmax sftlf sftof ta tas tasmax tasmin ts 
tsl ua uas va vas zg'
```

The QC files should now show errors related to the missing file(s), in addition to the QC error messages for the previously introduced bugs.
